pkg/controller/deployment/deployment_controller.go

```diff
@@ -22,68 +22,311 @@ import (
+	if d := dc.getDeploymentForRC(curRC); d != nil {
+		dc.enqueueDeployment(d)
+	}
+	// A number of things could affect the old deployment: labels changing,
+	// pod template changing, etc.
+	oldRC := old.(*api.ReplicationController)
+	// TODO: Is this the right way to check this, or is checking names sufficient?
+	if !api.Semantic.DeepEqual(oldRC, curRC) {
+		if oldD := dc.getDeploymentForRC(oldRC); oldD != nil {
+			dc.enqueueDeployment(oldD)
+		}
 	}
-	for _, deployment := range list.Items {
-		if err := d.reconcileDeployment(&deployment); err != nil {
-			return fmt.Errorf("error in reconciling deployment: %v", err)
+}
+
+// When a controller is deleted, enqueue the deployment that manages it.
+// obj could be an *api.ReplicationController, or a DeletionFinalStateUnknown
+// marker item.
+func (dc *DeploymentController) deleteRC(obj interface{}) {
+	rc, ok := obj.(*api.ReplicationController)
+
+	// When a delete is dropped, the relist will notice a pod in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the RC
+	// changed labels the new deployment will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
+		}
+		rc, ok = tombstone.Obj.(*api.ReplicationController)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not an rc %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
 		}
 	}
-	return nil
+	if d := dc.getDeploymentForRC(rc); d != nil {
+		dc.enqueueDeployment(d)
+	}
+}
+
+// obj could be an *api.Deployment, or a DeletionFinalStateUnknown marker item.
+func (dc *DeploymentController) enqueueDeployment(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	// TODO: Handle overlapping deployments better. Either disallow them at admission time or
+	// deterministically avoid syncing deployments that fight over RC's. Currently, we only
+	// ensure that the same deployment is synced for a given RC. When we periodically relist
+	// all deployments there will still be some RC instability. One way to handle this is
+	// by querying the store for all deployments that this deployment overlaps, as well as all
+	// deployments that overlap this deployments, and sorting them.
+	dc.queue.Add(key)
+}
+
+func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go dc.dController.Run(stopCh)
+	go dc.rcController.Run(stopCh)
+	go dc.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dc.worker, time.Second, stopCh)
```

Comment:
On Mon, Sep 21, 2015 at 1:50 PM, Sam Ghods <notifications@github.com> wrote:

> In pkg/controller/deployment/deployment_controller.go
> <https://github.com/kubernetes/kubernetes/pull/14130#discussion_r40023935>
> :
>
> > +	// TODO: Handle overlapping deployments better. Either disallow them at admission time or
> > +	// deterministically avoid syncing deployments that fight over RC's. Currently, we only
> > +	// ensure that the same deployment is synced for a given RC. When we periodically relist
> > +	// all deployments there will still be some RC instability. One way to handle this is
> > +	// by querying the store for all deployments that this deployment overlaps, as well as all
> > +	// deployments that overlap this deployments, and sorting them.
> > +	dc.queue.Add(key)
> > +}
> > +
> > +func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
> > +	defer util.HandleCrash()
> > +	go dc.dController.Run(stopCh)
> > +	go dc.rcController.Run(stopCh)
> > +	go dc.podController.Run(stopCh)
> > +	for i := 0; i < workers; i++ {
> > +		go util.Until(dc.worker, time.Second, stopCh)
>
> The current plan is to not block the deployment controller goroutines
> except while creating the desired RC
>
We dont really need to block while creating the RC. We can trigger a create
request and set the expectation and then wait till the expectation is
satisfied.

> and issuing calls for adjusting the replica counts. Instead, we will
> launch a timer to requeue the specific deployment controller at the
> estimated time when the next change should be made (and then in some
> interval beyond that if the pods are not ready yet). (N.B. this approach is
> not fully implemented yet in this PR.) For example, if scaleUp adds three
> replicas to the new RC, it will set a timer for an estimated time that the
> next scaleUp can happen based on the readiness probe information plus
> MinReadinessSeconds. If, after initial readiness probe time of the pod plus
> min readiness seconds, the pod is still not ready, it will continuously
> requeue every second(?) and reprocess the deployment.
>
> This means that we're constantly processing creating RC's and adjusting
> counts but never blocking the worker pool waiting for pods to be ready. The
> code becomes a little more complex than synchronously waiting, but it
> should be reasonably efficient because we cache all relevant state that we
> would query in memory locally in the informers (deployment, RC and pod
> objects)
>
replication controller manager has the same pattern.


> and more importantly it has the benefit that if the deployment controller
> were to crash (or in the alternate design, if a pod running a deployment
> were to crash) the deployments pick up right where they left off.
>
Wont the stores be reset, when controller crashes?


> cc @nikhiljindal <https://github.com/nikhiljindal> @bgrant0607
> <https://github.com/bgrant0607> for thoughts
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14130/files#r40023935>.
>